File: server/app/schemas/alerts.py
"""Alert response models (spec section 34)."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import AlertSeverity, AlertStatus, AlertType


class Alert(BaseModel):
    """One operational alert.

    Spec section 34 requires a title, description, timestamp, severity and the
    related machine or component. Section 45 requires that severity is readable
    as text rather than conveyed by colour, which is why `severity_label` and
    `description` are both mandatory and non-empty.
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    alert_type: AlertType
    alert_type_label: str = Field(description="Human-readable alert category.")
    severity: AlertSeverity
    severity_label: str = Field(description="Info, Warning or Critical, as display text.")
    status: AlertStatus
    status_label: str

    title: str = Field(min_length=1)
    description: str = Field(min_length=1, description="Full sentence describing the condition.")

    machine_id: UUID | None = None
    machine_code: str | None = None
    machine_name: str | None = None
    component_id: UUID | None = None
    component_name: str | None = None
    inventory_item_id: UUID | None = None
    inventory_item_name: str | None = None
    line_id: UUID | None = None
    line_name: str | None = None

    triggered_at: datetime = Field(description="When the condition was detected (UTC).")
    acknowledged_at: datetime | None = None
    resolved_at: datetime | None = None
    age_minutes: int = Field(ge=0, description="Minutes since the alert was triggered.")


class AlertSeverityCount(BaseModel):
    """How many open alerts sit at one severity."""

    severity: AlertSeverity
    severity_label: str
    count: int = Field(ge=0)


class AlertSummary(BaseModel):
    """Counts for the dashboard alerts panel."""

    total_open: int = Field(ge=0)
    critical_count: int = Field(ge=0)
    warning_count: int = Field(ge=0)
    info_count: int = Field(ge=0)
    acknowledged_count: int = Field(ge=0)
    severity_breakdown: list[AlertSeverityCount] = Field(default_factory=list)

File: client/types/alerts.ts
/**
 * Alert response models, mirroring `app/schemas/alerts.py`.
 *
 * Every alert carries a title, a full-sentence description and a severity
 * label. Spec section 34 requires all three, and spec section 45 requires that
 * severity is readable as text rather than conveyed by a colour.
 */
import type {
  AlertSeverity,
  AlertStatus,
  AlertType,
  IsoDateTime,
  PaginationFilters,
  Uuid,
} from "@/types/common";

/** One operational alert. */
export interface Alert {
  id: Uuid;
  alert_type: AlertType;
  alert_type_label: string;
  severity: AlertSeverity;
  severity_label: string;
  status: AlertStatus;
  status_label: string;

  title: string;
  /** A full sentence. Never empty; a database constraint enforces it. */
  description: string;

  /** At least one subject is always set. */
  machine_id: Uuid | null;
  machine_code: string | null;
  machine_name: string | null;
  component_id: Uuid | null;
  component_name: string | null;
  inventory_item_id: Uuid | null;
  inventory_item_name: string | null;
  line_id: Uuid | null;
  line_name: string | null;

  triggered_at: IsoDateTime;
  acknowledged_at: IsoDateTime | null;
  resolved_at: IsoDateTime | null;
  age_minutes: number;
}

/** How many open alerts sit at one severity. */
export interface AlertSeverityCount {
  severity: AlertSeverity;
  severity_label: string;
  count: number;
}

/** Counts for the dashboard alerts panel. */
export interface AlertSummary {
  total_open: number;
  critical_count: number;
  warning_count: number;
  info_count: number;
  acknowledged_count: number;
  severity_breakdown: AlertSeverityCount[];
}

/** Filters accepted by the alerts list. */
export interface AlertFilters extends PaginationFilters {
  status?: AlertStatus;
  severity?: AlertSeverity;
  machine_id?: Uuid;
}
